algopy_testing/tests/primitives/test_uint64.py

```diff
@@ -0,0 +1,267 @@
+import algokit_utils
+import algosdk
+import hypothesis.strategies as st
+import pytest
+from algokit_utils import get_localnet_default_account
+from algokit_utils.config import config
+from algopy import UInt64
+from algopy.error.uint64 import UInt64OverflowError, UInt64UnderflowError
+from algopy.primitives.constants import MAX_UINT64, MAX_UINT64_BIT_SHIFT
+from algosdk.v2client.algod import AlgodClient
+from algosdk.v2client.indexer import IndexerClient
+from hypothesis import example, given, settings
+
+from tests.artifacts.PrimitiveOps.client import PrimitiveOpsContractClient
+
+EXAMPLES_PER_AVM_TEST = 5
+
+
+@pytest.fixture(scope="session")
+def primitive_ops_client(
+    algod_client: AlgodClient, indexer_client: IndexerClient
+) -> PrimitiveOpsContractClient:
+    """
+    Fixture for the PrimitiveOpsContractClient.
+    """
+
+    config.configure(
+        debug=True,
+    )
+
+    client = PrimitiveOpsContractClient(
+        algod_client,
+        creator=get_localnet_default_account(algod_client),
+        indexer_client=indexer_client,
+    )
+
+    client.deploy(
+        on_schema_break=algokit_utils.OnSchemaBreak.AppendApp,
+        on_update=algokit_utils.OnUpdate.AppendApp,
+    )
+    return client
+
+
+@given(st.integers(min_value=MAX_UINT64 + 1, max_value=2 * MAX_UINT64))
+def test_uint64_overflow(value: int) -> None:
+    """
+    Test that the python implementation of UInt64 raise an error when
+    given a value greater than the maximum value for UInt64.
+
+    NOTE: this scenario does not evaluate against AVM since algod validated parameters prior
+    to passing for evaluation in AVM.
+    Hence, it will always throw encoding errors (in cases of algosdk) OR http errors in case
+    of direct invocation. Refer to other scenarios
+    testing conditions when an overflow can happen as a result of a dynamic computation with
+    in AVM.
+    """
+
+    # Test against the direct implementation
+    with pytest.raises(UInt64OverflowError):
+        UInt64(value)
+
+
+@given(st.integers(min_value=-2 * MAX_UINT64, max_value=-1))
+def test_uint64_underflow(value: int) -> None:
+    """Test that underflow exceptions are raised for invalid negative values."""
+    # Test against the direct implementation
+    with pytest.raises(UInt64UnderflowError):
+        UInt64(value)
+
+
+@given(
+    st.integers(min_value=0, max_value=MAX_UINT64), st.integers(min_value=0, max_value=MAX_UINT64)
+)
+@settings(max_examples=EXAMPLES_PER_AVM_TEST)
+def test_uint64_addition(primitive_ops_client: PrimitiveOpsContractClient, a: int, b: int) -> None:
+    """Test the addition operation for UInt64 instances."""
+    if a + b > MAX_UINT64:
```

Comment:
The way hypothesis works, this won't be guaranteed to be hit just because the range includes them.

I'm not sure hypothesis provides much advantage here, but if  it is used, we'll need to add specific examples for boundary conditions (hypothesis allows this with `@example`)

Reply:
For now have just used regular pytest parameters, have left hypothesis as a dev dependancy for usages in more complex scenarios